File: src/bun.js/api/crypto/PBKDF2.zig
const PBKDF2 = @This();

password: jsc.Node.StringOrBuffer = jsc.Node.StringOrBuffer.empty,
salt: jsc.Node.StringOrBuffer = jsc.Node.StringOrBuffer.empty,
iteration_count: u32 = 1,
length: i32 = 0,
algorithm: EVP.Algorithm,

pub fn run(this: *PBKDF2, output: []u8) bool {
    const password = this.password.slice();
    const salt = this.salt.slice();
    const algorithm = this.algorithm;
    const iteration_count = this.iteration_count;
    const length = this.length;

    @memset(output, 0);
    assert(this.length <= @as(i32, @intCast(output.len)));
    BoringSSL.ERR_clear_error();
    const rc = BoringSSL.PKCS5_PBKDF2_HMAC(
        if (password.len > 0) password.ptr else null,
        @intCast(password.len),
        salt.ptr,
        @intCast(salt.len),
        @intCast(iteration_count),
        algorithm.md().?,
        @intCast(length),
        output.ptr,
    );

    if (rc <= 0) {
        return false;
    }

    return true;
}

pub const Job = struct {
    pbkdf2: PBKDF2,
    output: []u8 = &[_]u8{},
    task: jsc.WorkPoolTask = .{ .callback = &runTask },
    promise: jsc.JSPromise.Strong = .{},
    vm: *jsc.VirtualMachine,
    err: ?u32 = null,
    any_task: jsc.AnyTask = undefined,
    poll: Async.KeepAlive = .{},

    pub const new = bun.TrivialNew(@This());

    pub fn runTask(task: *jsc.WorkPoolTask) void {
        const job: *PBKDF2.Job = @fieldParentPtr("task", task);
        defer job.vm.enqueueTaskConcurrent(jsc.ConcurrentTask.create(job.any_task.task()));
        job.output = bun.default_allocator.alloc(u8, @as(usize, @intCast(job.pbkdf2.length))) catch {
            job.err = BoringSSL.EVP_R_MEMORY_LIMIT_EXCEEDED;
            return;
        };
        if (!job.pbkdf2.run(job.output)) {
            job.err = BoringSSL.ERR_get_error();
            BoringSSL.ERR_clear_error();

            bun.default_allocator.free(job.output);
            job.output = &[_]u8{};
        }
    }

    pub fn runFromJS(this: *Job) bun.JSTerminated!void {
        defer this.deinit();
        if (this.vm.isShuttingDown()) {
            return;
        }

        const globalThis = this.vm.global;
        const promise = this.promise.swap();
        if (this.err) |err| {
            try promise.reject(globalThis, createCryptoError(globalThis, err));
            return;
        }

        const output_slice = this.output;
        assert(output_slice.len == @as(usize, @intCast(this.pbkdf2.length)));
        const buffer_value = jsc.JSValue.createBuffer(globalThis, output_slice);
        this.output = &[_]u8{};
        try promise.resolve(globalThis, buffer_value);
    }

    pub fn deinit(this: *Job) void {
        this.poll.unref(this.vm);
        this.pbkdf2.deinitAndUnprotect();
        this.promise.deinit();
        bun.default_allocator.free(this.output);
        bun.destroy(this);
    }

    pub fn create(vm: *jsc.VirtualMachine, globalThis: *jsc.JSGlobalObject, data: *const PBKDF2) *Job {
        var job = Job.new(.{
            .pbkdf2 = data.*,
            .vm = vm,
            .any_task = undefined,
        });

        job.promise = jsc.JSPromise.Strong.init(globalThis);
        job.any_task = jsc.AnyTask.New(@This(), &runFromJS).init(job);
        job.poll.ref(vm);
        jsc.WorkPool.schedule(&job.task);

        return job;
    }
};

pub fn deinitAndUnprotect(this: *PBKDF2) void {
    this.password.deinitAndUnprotect();
    this.salt.deinitAndUnprotect();
}

pub fn deinit(this: *PBKDF2) void {
    this.password.deinit();
    this.salt.deinit();
}

pub fn fromJS(globalThis: *jsc.JSGlobalObject, callFrame: *jsc.CallFrame, is_async: bool) bun.JSError!PBKDF2 {
    const arg0, const arg1, const arg2, const arg3, const arg4, const arg5 = callFrame.argumentsAsArray(6);

    if (!arg3.isNumber()) {
        return globalThis.throwInvalidArgumentTypeValue("keylen", "number", arg3);
    }

    const keylen_num = arg3.asNumber();

    if (std.math.isInf(keylen_num) or std.math.isNan(keylen_num)) {
        return globalThis.throwRangeError(keylen_num, .{
            .field_name = "keylen",
            .msg = "an integer",
        });
    }

    if (keylen_num < 0 or keylen_num > std.math.maxInt(i32)) {
        return globalThis.throwRangeError(keylen_num, .{ .field_name = "keylen", .min = 0, .max = std.math.maxInt(i32) });
    }

    const keylen: i32 = @intFromFloat(keylen_num);

    if (globalThis.hasException()) {
        return error.JSError;
    }

    if (!arg2.isAnyInt()) {
        return globalThis.throwInvalidArgumentTypeValue("iterations", "number", arg2);
    }

    const iteration_count = try arg2.coerce(i64, globalThis);

    if (!globalThis.hasException() and (iteration_count < 1 or iteration_count > std.math.maxInt(i32))) {
        return globalThis.throwRangeError(iteration_count, .{ .field_name = "iterations", .min = 1, .max = std.math.maxInt(i32) + 1 });
    }

    if (globalThis.hasException()) {
        return error.JSError;
    }

    const algorithm = brk: {
        if (!arg4.isString()) {
            return globalThis.throwInvalidArgumentTypeValue("digest", "string", arg4);
        }

        invalid: {
            switch (try EVP.Algorithm.map.fromJSCaseInsensitive(globalThis, arg4) orelse break :invalid) {
                .shake128, .shake256, .@"sha3-224", .@"sha3-256", .@"sha3-384", .@"sha3-512" => break :invalid,
                else => |alg| break :brk alg,
            }
        }

        if (!globalThis.hasException()) {
            const slice = try arg4.toSlice(globalThis, bun.default_allocator);
            defer slice.deinit();
            const name = slice.slice();
            return globalThis.ERR(.CRYPTO_INVALID_DIGEST, "Invalid digest: {s}", .{name}).throw();
        }
        return error.JSError;
    };

    var out = PBKDF2{
        .iteration_count = @intCast(iteration_count),
        .length = keylen,
        .algorithm = algorithm,
    };
    defer {
        if (globalThis.hasException()) {
            if (is_async)
                out.deinitAndUnprotect()
            else
                out.deinit();
        }
    }

    const allow_string_object = true;
    out.salt = try jsc.Node.StringOrBuffer.fromJSMaybeAsync(globalThis, bun.default_allocator, arg1, is_async, allow_string_object) orelse {
        return globalThis.throwInvalidArgumentTypeValue("salt", "string or buffer", arg1);
    };

    if (out.salt.slice().len > std.math.maxInt(i32)) {
        return globalThis.throwInvalidArguments("salt is too long", .{});
    }

    out.password = try jsc.Node.StringOrBuffer.fromJSMaybeAsync(globalThis, bun.default_allocator, arg0, is_async, allow_string_object) orelse {
        return globalThis.throwInvalidArgumentTypeValue("password", "string or buffer", arg0);
    };

    if (out.password.slice().len > std.math.maxInt(i32)) {
        return globalThis.throwInvalidArguments("password is too long", .{});
    }

    if (is_async) {
        if (!arg5.isFunction()) {
            return globalThis.throwInvalidArgumentTypeValue("callback", "function", arg5);
        }
    }

    return out;
}

/// For usage in Zig
pub fn pbkdf2(
    output: []u8,
    password: []const u8,
    salt: []const u8,
    iteration_count: u32,
    algorithm: Algorithm,
) ?[]const u8 {
    var pbk = PBKDF2{
        .algorithm = algorithm,
        .password = jsc.Node.StringOrBuffer{ .encoded_slice = jsc.ZigString.Slice.fromUTF8NeverFree(password) },
        .salt = jsc.Node.StringOrBuffer{ .encoded_slice = jsc.ZigString.Slice.fromUTF8NeverFree(salt) },
        .iteration_count = iteration_count,
        .length = @intCast(output.len),
    };

    if (!pbk.run(output)) {
        return null;
    }

    return output;
}

const string = []const u8;

const std = @import("std");

const bun = @import("bun");
const Async = bun.Async;
const assert = bun.assert;
const default_allocator = bun.default_allocator;
const BoringSSL = bun.BoringSSL.c;

const jsc = bun.jsc;
const CallFrame = jsc.CallFrame;
const JSGlobalObject = jsc.JSGlobalObject;
const JSValue = jsc.JSValue;
const VirtualMachine = jsc.VirtualMachine;
const ZigString = jsc.ZigString;
const createCryptoError = jsc.API.Bun.Crypto.createCryptoError;

const EVP = jsc.API.Bun.Crypto.EVP;
const Algorithm = EVP.Algorithm;

File: src/runtime/crypto/PBKDF2.rs
use core::ffi::c_uint;

use bun_boringssl_sys as boringssl;
use bun_jsc::{
    ArrayBuffer, CallFrame, JSGlobalObject, JSValue, Job, JobContext, JsResult, JsThread, Strong,
};

use crate::node::{Flavor, StringObjects, StringOrBuffer, ThreadIsolated, ThreadIsolatedArg};

use crate::crypto::evp::{self, Algorithm};

pub(crate) struct PBKDF2 {
    pub password: StringOrBuffer<'static>,
    pub salt: StringOrBuffer<'static>,
    pub iteration_count: u32,
    pub length: usize,
    algorithm: Algorithm,
}
// SAFETY: `password` and `salt` are `StringOrBuffer`s (see its impl); the rest is plain data.
unsafe impl ThreadIsolatedArg for PBKDF2 {}

impl PBKDF2 {
    pub(crate) fn run(&mut self, output: &mut [u8]) -> bool {
        let password = self.password.slice();
        let salt = self.salt.slice();
        let algorithm = self.algorithm;
        let iteration_count = self.iteration_count;
        let length = self.length;

        output.fill(0);
        debug_assert!(self.length <= output.len());
        // Node.js (OpenSSL) rejects a zero-length derivation; BoringSSL accepts it.
        if length == 0 {
            return false;
        }
        boringssl::ERR_clear_error();
        // SAFETY: password/salt point to valid slices for the given lengths;
        // algorithm.md() returns a non-null EVP_MD; output is writable for `length` bytes.
        let rc = unsafe {
            boringssl::PKCS5_PBKDF2_HMAC(
                if !password.is_empty() {
                    password.as_ptr()
                } else {
                    core::ptr::null()
                },
                password.len(),
                salt.as_ptr(),
                salt.len(),
                iteration_count as c_uint,
                algorithm.md().unwrap(),
                length,
                output.as_mut_ptr(),
            )
        };

        if rc <= 0 {
            return false;
        }

        true
    }

    /// The second element is the validated callback on the `Async` flavor and
    /// `JSValue::UNDEFINED` on `Sync` (as `Scrypt::from_js`).
    pub(crate) fn from_js(
        global_this: &JSGlobalObject,
        call_frame: &CallFrame,
        flavor: Flavor,
    ) -> JsResult<(PBKDF2, JSValue)> {
        let [arg0, arg1, arg2, arg3, mut arg4, mut arg5] = call_frame.arguments_as_array::<6>();
        // pbkdf2(password, salt, iterations, keylen, callback): digest omitted.
        if flavor == Flavor::Async && arg4.is_function() {
            arg5 = arg4;
            arg4 = JSValue::UNDEFINED;
        }

        if !arg3.is_number() {
            return Err(global_this.throw_invalid_argument_type_value(b"keylen", b"number", arg3));
        }

        let keylen_num = arg3.as_number();

        if !arg3.is_integer() {
            return Err(global_this.throw_range_error(
                keylen_num,
                bun_jsc::RangeErrorOptions {
                    field_name: b"keylen",
                    msg: b"an integer",
                    ..Default::default()
                },
            ));
        }

        if keylen_num < 0.0 || keylen_num > i32::MAX as f64 {
            return Err(global_this.throw_range_error(
                keylen_num,
                bun_jsc::RangeErrorOptions {
                    field_name: b"keylen",
                    min: 0,
                    max: i32::MAX as i64,
                    ..Default::default()
                },
            ));
        }

        // 0..=i32::MAX was checked above.
        let keylen = keylen_num as usize;

        if !arg2.is_number() {
            return Err(global_this.throw_invalid_argument_type_value(
                b"iterations",
                b"number",
                arg2,
            ));
        }

        let iterations_num = arg2.as_number();

        if !arg2.is_integer() {
            return Err(global_this.throw_range_error(
                iterations_num,
                bun_jsc::RangeErrorOptions {
                    field_name: b"iterations",
                    msg: b"an integer",
                    ..Default::default()
                },
            ));
        }

        if iterations_num < 1.0 || iterations_num > i32::MAX as f64 {
            return Err(global_this.throw_range_error(
                iterations_num,
                bun_jsc::RangeErrorOptions {
                    field_name: b"iterations",
                    min: 1,
                    max: i32::MAX as i64,
                    ..Default::default()
                },
            ));
        }

        let iteration_count: i64 = iterations_num as i64;

        let algorithm = 'brk: {
            if !arg4.is_string() {
                return Err(
                    global_this.throw_invalid_argument_type_value(b"digest", b"string", arg4)
                );
            }

            'invalid: {
                let slice = arg4.to_utf8(global_this)?;
                match evp::lookup_ignore_case(slice.slice()) {
                    Some(alg) => match alg {
                        Algorithm::Shake128 | Algorithm::Shake256 => break 'invalid,
                        other if other.md().is_none() => break 'invalid,
                        other => break 'brk other,
                    },
                    None => break 'invalid,
                }
            }

            let slice = arg4.to_utf8(global_this)?;
            let name = slice.slice();
            return Err(global_this
                .err(
                    bun_jsc::ErrorCode::CRYPTO_INVALID_DIGEST,
                    format_args!("Invalid digest: {}", bstr::BStr::new(name)),
                )
                .throw());
        };

        let mut out = PBKDF2 {
            password: StringOrBuffer::default(),
            salt: StringOrBuffer::default(),
            iteration_count: u32::try_from(iteration_count).expect("int cast"),
            length: keylen,
            algorithm,
        };
        out.salt = match StringOrBuffer::from_js_maybe_async(
            global_this,
            arg1,
            flavor,
            StringObjects::Allow,
        )? {
            Some(v) => v,
            None => {
                return Err(global_this.throw_invalid_argument_type_value(
                    b"salt",
                    b"string or buffer",
                    arg1,
                ));
            }
        };

        if out.salt.slice().len() > i32::MAX as usize {
            return Err(global_this.throw_invalid_arguments(format_args!("salt is too long")));
        }

        out.password = match StringOrBuffer::from_js_maybe_async(
            global_this,
            arg0,
            flavor,
            StringObjects::Allow,
        )? {
            Some(v) => v,
            None => {
                return Err(global_this.throw_invalid_argument_type_value(
                    b"password",
                    b"string or buffer",
                    arg0,
                ));
            }
        };

        if out.password.slice().len() > i32::MAX as usize {
            return Err(global_this.throw_invalid_arguments(format_args!("password is too long")));
        }

        if flavor == Flavor::Sync {
            if let StringOrBuffer::Buffer(buffer) = &mut out.salt {
                buffer.buffer = ArrayBuffer::from_typed_array(global_this, buffer.buffer.value);
            }
        }

        let callback = match flavor {
            Flavor::Async => {
                if !arg5.is_function() {
                    return Err(global_this.throw_invalid_argument_type_value(
                        b"callback",
                        b"function",
                        arg5,
                    ));
                }
                arg5
            }
            Flavor::Sync => JSValue::UNDEFINED,
        };

        Ok((out, callback))
    }

    /// [`from_js`](Self::from_js) for the work-pool job, with its validated callback.
    pub(crate) fn from_js_async(
        global_this: &JSGlobalObject,
        call_frame: &CallFrame,
    ) -> JsResult<(ThreadIsolated<PBKDF2>, JSValue)> {
        let (data, callback) = Self::from_js(global_this, call_frame, Flavor::Async)?;
        // SAFETY: parsed with `Flavor::Async`.
        Ok((unsafe { ThreadIsolated::new(data) }, callback))
    }
}

/// `crypto.pbkdf2` off the JS thread.
pub(crate) struct Pbkdf2Job {
    pub pbkdf2: ThreadIsolated<PBKDF2>,
    pub output: Vec<u8>,
    pub err: bool,
}

/// JS-thread state for [`Pbkdf2Job`]: the user callback, invoked as `(err)` or `(null, buffer)`.
#[derive(bun_jsc::JsAffine)]
pub(crate) struct Pbkdf2Js {
    pub callback: Strong,
}

impl JobContext for Pbkdf2Job {
    type OffThread = Self;
    type Js = Pbkdf2Js;

    fn run(this: &mut Self, done: bun_jsc::Completion<Self>) -> Option<bun_jsc::Completion<Self>> {
        let len = this.pbkdf2.length;
        // `Vec` allocation aborts on OOM; use try_reserve to surface an error instead.
        let mut buf = Vec::new();
        if buf.try_reserve_exact(len).is_err() {
            this.err = true;
            return Some(done);
        }
        buf.resize(len, 0);
        this.output = buf;

        if !this.pbkdf2.run(&mut this.output) {
            this.err = true;
            boringssl::ERR_clear_error();
            this.output = Vec::new();
        }
        Some(done)
    }

    fn then(mut this: Self, js: Pbkdf2Js, cx: &JsThread<'_>) -> JsResult<()> {
        let global_this = cx.global();
        let event_loop = global_this.bun_vm().event_loop_mut();
        let callback = js.callback.get();
        if this.err {
            let err = global_this.create_error_instance(format_args!("PBKDF2 derivation failed"));
            event_loop.run_callback(callback, global_this, JSValue::UNDEFINED, &[err]);
            return Ok(());
        }

        let output_slice = core::mem::take(&mut this.output);
        debug_assert!(output_slice.len() == this.pbkdf2.length);
        // Ownership transfers to JSC (freed via MarkedArrayBuffer_deallocator → mimalloc free).
        match JSValue::create_buffer(global_this, output_slice.leak()) {
            Ok(buffer_value) => event_loop.run_callback(
                callback,
                global_this,
                JSValue::UNDEFINED,
                &[JSValue::NULL, buffer_value],
            ),
            // The result could not be built (allocation failure): that is this
            // derivation's error.
            Err(err) => event_loop.run_callback(
                callback,
                global_this,
                JSValue::UNDEFINED,
                &[global_this.take_error(err)],
            ),
        }
        Ok(())
    }
}

/// Schedule the derivation on the work pool; `callback` was validated by `from_js_async`.
pub(crate) fn create_job(
    global_this: &JSGlobalObject,
    data: ThreadIsolated<PBKDF2>,
    callback: JSValue,
) {
    let cx = global_this.js_thread();
    Job::<Pbkdf2Job>::schedule(
        &cx,
        Pbkdf2Job {
            pbkdf2: data,
            output: Vec::new(),
            err: false,
        },
        Pbkdf2Js {
            callback: Strong::create(
                callback.with_async_context_if_needed(global_this),
                global_this,
            ),
        },
    );
}
